Prevent toogle to false at restart of ADS platforms

carstenschroeder opened the pull request:
## Description:
Fixes behavior of ADS binary_sensor, light and switch entities to toggle to false at HA restart, while the underlying device's state kept being true.

**Related issue (if applicable):** fixes #21893

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

carstenschroeder committed:
Prevent toogle to false at restart

carstenschroeder committed:
add available property

carstenschroeder committed:
get initial state of device

carstenschroeder committed:
fix pylint

carstenschroeder committed:
fix pylint

carstenschroeder committed:
remove additional poll

carstenschroeder committed:
Forward state None in switch/light and binary sensor

carstenschroeder committed:
remove core changes

carstenschroeder committed:
add wait for initial stete update

pvizeli reviewed (changes requested):
You can't use threads/events in a async context. Maybe you solve your problem with `add_devices([..], true)`?

carstenschroeder committed:
use add_entities([], True)

carstenschroeder committed:
use add_entities([], True)

carstenschroeder commented:
@pvizeli Thanks for your help! Do you mean the way I did it in the last commit for binary_sensor.py?

carstenschroeder committed:
add availability

carstenschroeder committed:
add availability

carstenschroeder committed:
convert to asyncio.Event w/ async_timeout.timeout

carstenschroeder committed:
convert to asyncio.Event w/ async_timeout.timeout

carstenschroeder committed:
fix houndci findings

carstenschroeder committed:
fix houndci findings

carstenschroeder committed:
Fix houndci findings II

carstenschroeder committed:
Fix houndci findings II

carstenschroeder commented:
@pvizeli With some help in discord I found a solution that uses `asyncio.Event` w/ `async_timeout.timeout`. This does exactly what I want w/o blocking the event loop. Could you please have a look?

carstenschroeder committed:
make event.set() treadsafe

carstenschroeder committed:
make event.set() treadsafe

balloob committed:
Bumped version to 0.89.0b0

balloob committed:
Version bump to 0.90.0b0

robbiet480 committed:
Mobile App: Require encryption for registrations that support it (#21852)

## Description:

**Related issue (if applicable):** fixes #21758

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

robbiet480 committed:
Mobile App: Support rendering multiple templates at once (#21851)

* Support rendering multiple templates at once

* Only catch TemplateError and dont log the error

robbiet480 committed:
Mobile App: Update Location schema updates & device ID generation (#21849)

* Update location schema

* Generate a random device ID at registration time for later use with device_tracker.see

* Remove host name from device_tracker.see payload

* Drop consider_home from the payload

* Remove stale consider_home in schema

* Remove source_type

philhawthorne committed:
Remove UTF8 decoding for Waze (#22020)

Removes the UFT8 decoding for the Waze sensor, which broke in 0.89

Fixes #21739

robbiet480 committed:
Mobile App: Enable loading via discovery (surprise inside!) (#22027)

![](http://funpeep.com/wp-content/uploads/2014/04/Cute-White-Cat-Wallpaper.jpg)

robbiet480 committed:
Mobile App: Discovery to default configuration.yaml, zeroconf to default_config (#22028)

* Move discovery into default configuration.yaml

* Add zeroconf to default_config

JeffLIrion committed:
Bump androidtv to 0.0.11 (#22025)

mvn23 committed:
Bump pyotgw to 0.4b2 (#21973)

awarecan committed:
Fix lifx light async error (#22031)

carstenschroeder committed:
use await async_add_executor_job

carstenschroeder committed:
use await async_add_executor_job

cgtobi committed:
Bring back the boiler status (#22021)

emontnemery committed:
Update additional platforms to use new MQTT message callback (#22030)

* Move additional platforms to new MQTT callback

* Fix automation.mqtt

robbiet480 committed:
Mobile App: Register devices into the registry (#21856)

* Register devices into the registry

* Switch to device ID instead of webhook ID

* Rearchitect mobile_app to support config entries

* Kill DATA_REGISTRATIONS by migrating registrations into config entries

* Fix tests

* Improve how we get the config_entry_id

* Remove single_instance_allowed

* Simplify setup_registration

* Move webhook registering functions into __init__.py since they are only ever used once

* Kill get_registration websocket command

* Support description_placeholders in async_abort

* Add link to mobile_app implementing apps in abort dialog

* Store config entry and device registry entry in hass.data instead of looking it up

* Add testing to ensure that the config entry is created at registration

* Fix busted async_abort test

* Remove unnecessary check for entry is None

pvizeli committed:
Fix Google Assistant User with Cloud (#22042)

* Fix Google Assistant User with Cloud

* Fix User Agent ID

* respell

* Fix object

* Fix tests

* fix lint

* Fix lint

balloob committed:
Bumped version to 0.90.0b1

robbiet480 committed:
Mobile App: Sensors (#21854)

## Description:

**Related issue (if applicable):** fixes #21782

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

robbiet480 committed:
Mobile App: Expose Cloud Remote UI FQDN in registration response (#22055)

* Add a callback to get the cloud remote UI FQDN

* Expose Cloud Remote UI FQDN in the registration response

* Return a URL instead of FQDN

balloob committed:
Return config entry ID after creation (#22060)

balloob committed:
Fix test

balloob committed:
Fix more test

awarecan committed:
Correct context (#22061)

pvizeli committed:
Binary Sensor for Remote UI & Fix timezone (#22076)

* Binary Sensor for Remote UI

* Fix lint

* Revert make hass public

* Add tests

balloob committed:
Updated frontend to 20190315.0

balloob committed:
Bumped version to 0.90.0b2

balloob committed:
Fix func

JeffLIrion committed:
Bump androidtv to 0.0.12 (#22072)

balloob committed:
Prevent cloud remote UI when using 127.0.0.1 as trusted network (#22093)

* Prevent cloud remote UI when using trusted networks

* Limit to 127.0.0.1 trusted network

* Update error msg

* Disable ipv6 loopback

balloob committed:
Updated frontend to 20190315.1

hunterjm committed:
Beta Fix: FFMPEG and Stream component (#22091)

* remove stream_source from ffmpeg and onvif and add to generic ip cam

* fix tests

balloob committed:
Bumped version to 0.90.0b3

mvn23 committed:
Fix opentherm_gw blocks HA startup when gateway unreachable. (#22106)

mvn23 committed:
Fix TypeError in current_temperature if no temperature is known. (#22112)

Don't set opentherm_gw climate temperatures to 0 on init.

mvn23 committed:
Improve opentherm gw startup (#22121)

* Improve fix in c90f0d5 (#22106). Schedule connect coroutine directly on the loop rather than waiting for EVENT_HOMEASSISTANT_START.

* Remove unused import.

hunterjm committed:
delete previously removed service option from services yaml (#22123)

balloob committed:
Updated frontend to 20190316.0

balloob committed:
Bumped version to 0.90.0b4

awarecan committed:
Remove hass.config from aws_lambda notify payload (#22125)

balloob committed:
Allow non-admins to listen to certain events (#22137)

WebSpider committed:
Bump tado version (#22145)

* Bump python-tado, new API endpoint

* Change references of old API endpoint to new

* Update REQUIREMENTS

hunterjm committed:
Fix resetting access token on streams with keepalive (#22148)

balloob committed:
Updated frontend to 20190318.0

frenck committed:
Upgrade toonapilib to 3.2.2 + lower interval (#22160)

pvizeli committed:
Remove config check over supervisor (#22156)

* Remove config check over supervisor

* Fix lint

* Fix tests

balloob committed:
Bumped version to 0.90.0b5

uchagani committed:
bump total_connect_client to 0.24 (#22166)

pvizeli commented:
Can you explain the program flow they you try to solve with this solution?

carstenschroeder commented:
With this PR I try to address an general issue of HA during restart. Especially for push components the entity state changes for a short period of time to e.g. unknown state or false. Depending how the platform code handles it. The root cause is, that `_async_add_entity` reads the state of the new entity (e.g. binary_sensor) at a point in time, when the platform has not yet received the current state from the device. So the state has a initial value. (see also issue #21893)

In order to achieve a clean state graph and to prevent automations to be trigged, I would like to prevent that state change for the ADS component. 

My solution is just that method `async_added_to_hass` waits in a async friendly way and with a timeout for the device callback which delivers the actual state. So that _async_add_entity will read the current state of the device and not a random initial state value.

carstenschroeder commented:
Components with `should_poll == True` can achieve the same by using `add_devices([..], true)`, so they are able to provide a valid state in `update()`. But this is not an option here.

pvizeli committed:
Update Hass-NabuCasa 0.8 (#22177)

balloob committed:
Always load Hass.io component on Hass.io (#22185)

* Always load Hass.io component on Hass.io

* Lint

* Lint

balloob committed:
Updated frontend to 20190319.0

balloob committed:
Bumped version to 0.90.0b6

balloob committed:
Updated frontend to 20190319.1

balloob committed:
Bumped version to 0.90.0b7

balloob committed:
Updated frontend to 20190320.0

Swamp-Ig committed:
Fixed typing errors (#22207)

balloob committed:
Bumped version to 0.90.0

balloob committed:
Merge pull request #22216 from home-assistant/rc

0.90.0

Adminiuga committed:
Fix ZHA force polled entities. (#22222)

## Description:
Fix "force_polled" ZHA entities.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

carstenschroeder committed:
Merge branch 'dev' into ads-binary-sensor-fix

hunterjm committed:
Stream fixes (#22238)

* fix issues with out of order packets, and empty first packet on some IP camera models

* do not skip the first packet

hunterjm committed:
Stream fixes (#22238)

* fix issues with out of order packets, and empty first packet on some IP camera models

* do not skip the first packet

awarecan committed:
Fix build issue (#22251)

Adminiuga committed:
add ZHA channel name property (#22218)

* Make channel name a property.

* Cleanup Zigbee channels.

Use zcl.Cluster.ep_attribute as default channel name.

michaelarnauts committed:
Allow on/off on tado climate component. (#22242)

* Allow on/off on tado climate component. Bump python-tado version. Patch from @wmalgadey

* Revert wrongly change in tado device tracker

michaelarnauts committed:
Allow on/off on tado climate component. (#22242)

* Allow on/off on tado climate component. Bump python-tado version. Patch from @wmalgadey

* Revert wrongly change in tado device tracker

andrewsayre committed:
Fix validate webhook requirements (#22248)

andrewsayre committed:
Fix validate webhook requirements (#22248)

kkr16 committed:
Bump total-connect-client to 0.25, fixing issue with Total Connect (#22230)

* Bump total-connect-client to 0.25

* Bump version in requirements_all.txt

kkr16 committed:
Bump total-connect-client to 0.25, fixing issue with Total Connect (#22230)

* Bump total-connect-client to 0.25

* Bump version in requirements_all.txt

Swamp-Ig committed:
Handle on/off through TemperatrureSetting trait. (#21842)

pvizeli committed:
Update Hass-NabuCasa 0.9 (#22258)

pvizeli committed:
Update Hass-NabuCasa 0.9 (#22258)

balloob committed:
Updated frontend to 20190321.0

balloob committed:
Updated frontend to 20190321.0

balloob committed:
Update translations

rdbahm committed:
Change .now() to .utcnow() (#22233)

ephem works in UTC time, not local time. As a result, our comparison needs to be in UTC for accurate results.

balloob committed:
Bumped version to 0.90.1

balloob committed:
Lint

Danielhiversen committed:
Upgrade tibber libary, support solar production (#22261)

pvizeli committed:
Update hass-nabucasa 0.10 (#22267)

pvizeli committed:
Update hass-nabucasa 0.10 (#22267)

balloob committed:
Merge pull request #22265 from home-assistant/rc

0.90.1

fabaff committed:
Upgrade sqlalchemy to 1.3.0 (#22269)

jjlawren committed:
Plex: Avoid refreshing by both device and session methods (#22266)

* Avoid refreshing by both device and session methods

* Fix indentation

JeffLIrion committed:
Bump androidtv to 0.0.13 (#22279)

* Bump androidtv to 0.0.13

* Bump androidtv to 0.0.13 in requirements_all.txt

JeffLIrion committed:
Improved exception handling and logging (#22268)

Catch RuntimeError exceptions. Don't log ADB command output if there isn't any.

amelchio committed:
Consolidate the netgear_lte configuration (#22105)

* Consolidate the netgear_lte configuration

* Simplfications from review

* Extract sensor_types

* Simplify defaults

Swamp-Ig committed:
Fixing the api_streams sensor (#22200)

* Fire events with websocket messages.

* Added tests to validate

* Fixed api_streams sensor to use new sensor

* Delete from coverageac as now works.

* Removed websocket request event.

* Use dispatcher instead of events.

* Moved sensor to under websocket_api

* Changes as per code review

* Fixed tests.

* Modified test

* Patch

jjlawren committed:
Fix progress for Plex media_players (#22224)

* Return current position and last updated timestamp needed by UI

* Add throttling to position updating

* Simplify logic, don't clear position when refreshing

* Use seconds

* Update homeassistant/components/plex/media_player.py

Co-Authored-By: jjlawren <jjlawren@users.noreply.github.com>

* Add throttling to position updating

* Simplify logic, don't clear position when refreshing

ljmerza committed:
Add sort by config and tests for Reddit integration (#22081)

* add sort type to reddit sensor

add sort type to reddit sensor

automated commit 14/03/2019 15:53:15

automated commit 15/03/2019 09:43:39

reddit sensor add tests and sort by config

reddit sensor add tests and sort by config

reddit sensor add tests and sort by config

reddit sensor add tests and sort by config

* Add reddit sensor sort_by config and tests

* Add reddit sensor sort_by config and tests

* Add reddit sensor sort_by config and tests

* Add reddit sensor sort_by config and tests

* Add reddit sensor sort_by config and tests

robbiet480 committed:
Fix Prometheus casting issues (#22282)

## Description:

Fix Prometheus casting issues

**Related issue (if applicable):** fixes #8659.
## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

Swamp-Ig committed:
Update trait to support auto without ranges. (#21847)

cgtobi committed:
Sort code owners alphabetically (#22304)

## Description:
Sort the code oweners list alphabetically.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

robbiet480 committed:
Update srpenergy library (#22307)

Fixes #18899

robbiet480 committed:
Switch from using Google Maps API for elevation to Open Elevation API (#22306)

## Description:

Switches elevation helper to use [Open Elevation](https://open-elevation.com/) instead of Google Maps API which now requires a API key. It's a drop in replacement for Google Maps too!

**Related issue (if applicable):** fixes #19860

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

robbiet480 committed:
Fix for embedded MQTT server configuration (#22305)

## Description:

Passing in a configuration for the embedded MQTT server has been broken for a while. This fixes that. See related issue number for further details.

**Related issue (if applicable):** fixes #18228

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

fabaff committed:
Upgrade pylast to 3.1.0 (#22302)

fabaff committed:
Upgrade py-cpuinfo to 5.0.0 (#22287)

fabaff committed:
Update abbreviation (#22317)

Commit:
Add homematicip cloud connection quality related attributes (#21990)

SNoof85 committed:
Add Freebox switch platform (#21710)

* Added Freebox switch and bump aiofreepybox version

* Added missing modified files

* removed unused import

* gen_requirements_all passed

* removed unused import

* Remove unused code

* lint fixes

* More lint fixe

* Bump aiofreepybox version and API version to Freebox

* Remove URL from log entry

* import relative

* Sort imports

balloob committed:
Google Assistant: Add camera stream trait (#22278)

* Add camera stream trait

* Lint

thibmaek committed:
Remove occupancy as it is not available in API (#22320)

fabaff committed:
Update file header (#22318)

* Update file header

* Fix indent

* Fix lint issue

balloob committed:
Consolidate more platforms (#22308)

* Consolidate final platforms

* Fix some tests

* Fix more tests

* Fix more tests

hunterjm committed:
fix where PLATFORM_SCHEMA gets pulled from (#22334)

uchagani committed:
show which component is causing translation errors (#22340)

renemarc committed:
Define GTFS sensor as a timestamp device class (#21053)

zewelor committed:
Move yeelight into component (#21593)

Kane610 committed:
Axis config flow (#18543)

* Initial draft

* Add tests for init
Fix hound comments

* Add tests for device
Change parameter handling to make device easier to test

* Remove superfluous functionality per Martins request

* Fix hound comments

* Embedded platforms

* Fix device import

* Config flow retry

* Options default values will be set automatically to options in config entry before component can be used

* Clean up init
Add populate options
Fix small issues in config flow
Add tests covering init

* Improve device tests

* Add config flow tests

* Fix hound comments

* Rebase miss

* Initial tests for binary sensors

* Clean up
More binary sensor tests

* Hound comments

* Add camera tests

* Fix initial state of sensors

* Bump dependency to v17

* Fix pylint and flake8

* Fix comments

MatthewFlamm committed:
Fix pressure in dark sky and openweathermap and add pressure utility (#21210)

GuryYu committed:
Fix xiaomi aqara cube with lumi.acpartner.v3 gateway (#22130)

Kane610 committed:
Support deCONZ library with exception handling (#21952)

amelchio committed:
Do not warn when creating an empty database (#22343)

syssi committed:
Add color support to Philips Moonlight (#22204)

syssi committed:
Add support for the power socket of the Xiaomi AC Partner V3 (#22205)

ktnrg45 committed:
Fix ps4 no creds with additional device (#22300)

* Fix no creds with additional device.

* Update config_flow.py

ktnrg45 committed:
Fix ps4 no creds with additional device (#22300)

* Fix no creds with additional device.

* Update config_flow.py

hunterjm committed:
Prefer TCP for RTSP streams (#22338)

## Description:

For RTSP streams, set the `prefer_tcp` FFMPEG flag.  This should resolve some of the "green feed" issues that some users are reporting, likely due to packets being lost over UDP on their network.

Resources: [FFMPEG protocols documentation](https://ffmpeg.org/ffmpeg-protocols.html#rtsp)

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

hunterjm committed:
Prefer TCP for RTSP streams (#22338)

## Description:

For RTSP streams, set the `prefer_tcp` FFMPEG flag.  This should resolve some of the "green feed" issues that some users are reporting, likely due to packets being lost over UDP on their network.

Resources: [FFMPEG protocols documentation](https://ffmpeg.org/ffmpeg-protocols.html#rtsp)

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

robbiet480 committed:
Set Onkyo reset log to debug instead of info (#22369)

Onkyo logs this message somewhat frequently, and its spammy, so lets make it a debug message instead of info. 

See also: #20081

cgtobi committed:
Upgrade netatmo smart_home module (#22365)

shanbs committed:
Fix the crash due to absence of the "default_home" in HomeData from pyatmo (netatmo/climate) (#22363)

StevenLooman committed:
Add support for 'image' media type (#22353)

hmmbob committed:
Warn user about HTML5 GCM deprecation (#22351)

* Warn user about GCM deprecation

* Fixing hound

* Fixing typo

* Fixing Travis fail

dilruacs committed:
Turn Panasonic Viera TV on without WOL (#22084)

* Turn the TV on via remote

* Turn the TV on via remote

* Use turn_on() from panasonic-viera==0.3.2

* make power option configurable

* add app_power as argument

* formatting

StevenLooman committed:
Properly connect sensors to hub (#21414)

* Properly connect sensors to hub
Refs #20958

* Don't connect (merge) with main device

* Provide manufacturer

* Linting

* Do connect upnp-sensors to main device

* Linting

* Fix requirements_all.txt

fredrike committed:
Add support for Tfiac Climate component (#21823)

## Description:

Add support for AC-models that follows the Tfiac protocol. Built together with @mellado.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8910

## Example entry for `configuration.yaml` (if applicable):
```yaml
climate:
   platform: tfiac
   host: 192.168.10.26
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23


Co-authored-by: Robbie Trencheny <me@robbiet.us>

awarecan committed:
Add aws component and consolidate aws notify platform  (#22240)

* Add aws component

* Move notify config under aws component

* Add basic tests for aws component

* Add deprecated warning for notify.aws_*

* Add more tests

karlkar committed:
Check if mac is set when more than 2 gateways (#21834)

* Check if mac is set when more than 2 gateways

When more than 2 gateways mac is required for each of them. Now voluptuous will require it.

* fix line length

* remove trailing whitespace

* Make it more readable

Mofeywalker committed:
Add 'method' parameter to forgiving_round method (#21708)

* Add 'method' parameter to forgiving_round method

Fixes #21707

* fix rounding behavior in round() filter

* add test cases for new rounding behaviour

nhorvath committed:
Expose detailed Ecobee equipment status (#20767)

* ecobee: expose detailed equipment status

* Fix #18244 for ecobee by moving current_operation property to current_operation_mode which is more accurate and defining current_operation properly, thanks @ZetaPhoenix

* fix docstring and lint issue

* Revert "fix docstring and lint issue"

This reverts commit d3a645f075f8a4017756f5eae05f00f05e2556cf.

* Revert "Fix #18244 for ecobee by moving current_operation property to current_operation_mode which is more accurate and defining current_operation properly, thanks @ZetaPhoenix"

This reverts commit bfd90551ef759b2c160d69da0778df89251e156b.

fabtesta committed:
Add ClickSend "caller" option (#20780)

* removed "from" parameter from ClickSend API call

Removed "from" parameter from API call to let ClickSend choose a random number and be more compliant with some cellular networks that do not allow incoming calls from the same recipient number.

* fixed "line too long (94 > 79 characters)" houndci code review

fixed "line too long (94 > 79 characters)" houndci code review

* Added new component optional parameter "caller".
If defined is used to override default logic that uses recipient phone number in "from" API call parameter

* Removed default value for CALLER parameter. If not defined then will take the RECIPIENT value.

Jc2k committed:
Give HomeKit locks better names by default (#22333)

## Description:

This is a follow up to #22171. There we set the name of an entity based on the `accessory-information` homekit service, rather than using the zeroconf/avahi name metadata. Unfortunately Lock also sets its name from zeroconf directly, rather than picking it up from the base class. This test updates it to be like the other homekit entities and use the base class.

(This is from my ongoing homekit_controller configentry branch).

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

zewelor committed:
Improve yeelight component (#22347)

ktnrg45 committed:
Improve PS4 media art fetching and config flow (#22167)

* improved config flow

* Added errors, docs url

* Added errors, docs url

* Added manual config mode

* Add tests for manual/auto host input

* fix inline docs

* fix inline docs

* Changed region list

* Added deprecated region message

* removed DEFAULT_REGION

* Added close method

* Fixes

* Update const.py

* Update const.py

* Update const.py

* Update test_config_flow.py

* Added invalid pin errors

* Update strings.json

* Update strings.json

* bump pyps4 to 0.5.0

* Bump pyps4 0.5.0

* Bump pyps4 to 0.5.0

* test fixes

* pylint

* Change error reference

* remove pin messages

* remove pin messages

* Update en.json

* remove pin tests

* fix tests

* update vol

* Vol fix

* Update config_flow.py

* Add migration for v1 entry

* lint

* fixes

* typo

* fix

* Update config_flow.py

* Fix vol

* Executor job for io method.

* Update __init__.py

* blank line

* Update __init__.py

* Update tests/components/ps4/test_config_flow.py

Co-Authored-By: ktnrg45 <38207570+ktnrg45@users.noreply.github.com>

JeffLIrion committed:
Allow for custom turn on/off commands (#22354)

Jc2k committed:
Remove unused const (#22383)

pvizeli committed:
Update hass-nabucasa & fix state (#22385)

* Update hass-nabucasa & fix state

* Fix lint

pvizeli committed:
Update hass-nabucasa & fix state (#22385)

* Update hass-nabucasa & fix state

* Fix lint

balloob committed:
Frontend indicate require admin (#22272)

* Allow panels to indicate they are meant for admins

* Panels to indicate when they require admin access

* Do not return admin-only panels to non-admin users

* Fix flake8

a005 committed:
Fix TpLink Device Tracker initialize error (#22349)

Catch RequestException instead of ConnectionError
In some cases TpLinkDeviceScanner throws various successors of
RequestException. They should be caught.

lapy committed:
Add traccar events (#22348)

* Add import_events(), small refactoring, traccar_id

I isolated the code that imports traccar tracking data and added a new function that imports the traccar events to make them run in parallel and reduce delay. The events that are imported in hass, will be fired with the prefix "traccar_". Furthermore a traccar_id is now imported in hass entities, useful for matching the traccar hass entities with the traccar hass events in the most accurate way.

* bump pytraccar version

* Code format fix

* Code format fix 2

* Code format fix 3

* Implement requested changes

* Add new traccar dependency

* Fix line too long

* Update device_tracker.py

* Update requirements_all.txt

zewelor committed:
Update python yeelight and add nightlight mode sensor (#22345)

fredrike committed:
updated pydaikin (#22382)

nmaggioni committed:
Support for Plex sensor with enforced SSL (#21432)

renemarc committed:
Search GTFS departures across midnight (#20992)

uchagani committed:
remove config sections from hass.config.components (#22370)

* remove config sections from hass.config.components

* fix tests

dshokouhi committed:
Update ring_doorbell to 0.2.3 (#22395)

Jc2k committed:
Fix unavailable state for homekit locks and covers (#22390)

nickw444 committed:
Add infer_arming_state option to ness alarm (#22379)

* Add infer_arming_state option to ness alarm

* actually use config value

* 🤦‍♂️

Kane610 committed:
deCONZ support Xiaomi vibration sensor (#22366)

* Martin pointed out in previous PR that no ending '.' in logging

* Add support for Xiaomi vibration sensor

syssi committed:
Fix data_key of the xiaomi_aqara cover for LAN protocol v2 (#22358)

Jc2k committed:
Remove get_serial helper that is no longer needed. (#22368)

awarecan committed:
Set default parallel_update value should base on async_update (#22149)

* Set default parallel_update value should base on async_update

* Set default parallel_update value should base on async_update

* Delay the parallel_update_semaphore creation

* Remove outdated comment

hunterjm committed:
reset unsub to None on timeout (#22404)

hunterjm committed:
reset unsub to None on timeout (#22404)

hunterjm committed:
Camera Preferences + Preload Stream (#22339)

* initial commit for camera preferences and preload stream

* cleanup and add tests

* respect camera preferences on each request stream call

* return the new prefs after update

balloob committed:
Move core stuff into Home Assistant integration (#22407)

* Move core stuff into Home Assistant integration

* Lint

Danielhiversen committed:
Tibber add support for Watty (#22397)

dmulcahey committed:
zha fixes (#22381)

zewelor committed:
Fix yeelight state update (#22373)

fronzbot committed:
Changed busy error to warning (#22398)

cgtobi committed:
Migrate tts (#22403)

* Migrate tts

* Migrate tts tests

* Update requirements

* Fix path to demo mp3

pvizeli committed:
Enable hass.io panel without ping (#22388)

* Enable hass.io panel without ping

* fix tests

pvizeli committed:
Enable hass.io panel without ping (#22388)

* Enable hass.io panel without ping

* fix tests

Jc2k committed:
Add missing append (#22414)

balloob committed:
Bumped version to 0.90.2

awarecan committed:
Update translate, fix dev build error (#22419)

amelchio committed:
Use dispatcher for netgear_lte state updates (#22328)

* Use dispatcher for netgear_lte state updates

* Also dispatch unavailable state

fbradyirl committed:
ciscomobilityexpress pypi version update (#22431)

* Bump pip

* Bump ciscomobilityexpress to 0.1.4

balloob committed:
Merge pull request #22420 from home-assistant/rc

0.90.2

pvizeli committed:
Update homeassistant-pyozw to 0.1.3 (#22433)

fbradyirl committed:
Adding conf for deep standby, wake and specific source bouquet of Enigma2 (#22393)

* - adding deep standby, conf for wake and specify source bouquet

* set defaults to strings

* bump pip

* bump pip

* bump pip

* bump pip

* bump pip

* bump pip

balloob committed:
Fix test name (#22421)

thomasloven committed:
Lower severity level of log messages from http.view (#21091)

pvizeli committed:
Update ha-ffmpeg 2.0 (#22427)

zewelor committed:
Add myself as codeowner for yeelight component (#22438)

Adminiuga committed:
Update ZHA component CODEOWNERS (#22452)

zewelor committed:
Add support for yeelight ceiling ambilight (#22346)

Swamp-Ig committed:
Bootstrap to start registry loading early (#22321)

* Registries store directly in data on loading.

* Loading registries concurent with stage 1.

* Removed comments

Kane610 committed:
Axis devices support device registry (#22367)

* Add support for device registry

* Fix test

ljmerza committed:
Use voluptuous error string for websocket validation error (#21883)

* use voluptuous error string to websocket validation error

* added exception logging to websocket error

* add detailed message to websocket validation error

* add error message to websocket validation error

* Add humanize error for websocket invalid vol error

* Add humanize error for websocket invalid vol error

* Add humanize error for websocket invalid vol error

rtclauss committed:
Add hourly forecasts to Dark Sky (#21820)

fredrike committed:
Add switches to control Daikin Airbase zones (#22417)

* initial AirBase zone support

* fix zone name

* version bump pydaikin

* don't use get()

heythisisnate committed:
better algorithm for computing unique_id (#22389)

GidoHakvoort committed:
Solaredge new sensors (#21047)

* Remove unused hass parameter for SolarEdgeData

* Add factory to create different types of sensors
Rename SolarEdgeSensor to SolarEdgeOverviewSensor
Rename SolarEdgeData to SolarEdgeOverviewDataService
Remove unused hass parameter in SolarEdgeOverviewDataService

* Add SolarEdgeDetailsDataService to retrieve details data
Add SolarEdgeDetailsSensor to report details data
Add abstract class SolarEdgeSensor
Add details sensor types

* Combine multiple details sensor into one sensor with attributes

* Fix pylint and flake8 errors

* Resolve conflict with solaredge component update

* Add SolarEdgeInventoryDataService to retrieve inventory information
Add SolarEdgeInventorySensor to view inventory information
Add inverters to monitored_conditions

* Fix pylint and flake8 errors

* Add additional monitored variables for solaredge

* Add new sensors to solaredge component

* Add SolarEdgePowerFlowDataService
Add SolarEdgePowerFlowSensor
Add new monitored_conditions for power consumption and grid, load and
solar power production/consumption

* Set entity_id for each sensor based on platform and sensor type

* Fix flake8 and pylint errors

* Add check for connections in return data

* Fix pylint and flake8 errors

* Renamed state_attributes to device_state_attributes
Moved request import to top

* Remove explicit definition of entity_id

* Fix pylint and flake8 errors

* Add check for None before adding sensor

* Update SolarEdgeSensorFactory with initial dict which maps sensor_key to
entity class and data service

* Update attribute values to snakecase
Added stingcase as requirement

* Update requirements_all.txt to include stringcase for solaredge

* Update some initial values for data and unit_of_measurement
Update sensor factory

awarecan committed:
Fix aws.notify platform schema (#22374)

* Fix aws component notify platform schema

* Address code review comment

* Do not allow load aws.notify from notify component

* Revert unrelated translation update

* Review comment

amelchio committed:
Do data extraction in sensors (#22444)

* Do data extraction in sensors

* Hopefully fix lint

Jc2k committed:
Remove homekit_controller duplicate legacy pairing loader code (#22442)

renemarc committed:
Cache GTFS metadata and expose utility attributes (breaking change) (#20966)

## Description:
Current sensor updates run 7 additional SQLite database queries to populate attributes, on top of the bus schedule queries themselves. Double that if you have two sensors. That leads to a lot of slowdowns for everything else when using an SD card!

Considering that some data never changes (agency, routes...) and that others like departure times are good until invalidated, let's fetch such metadata at first then only when relevant changes do occur.

**Breaking Change:**
GTFS sensor attributes are now named using the standard snake_case format.

### Work performed:
- All metadata queries are now cached.
- Metadata queries are now all regrouped in the `update()` method.
- Attributes assembling is now done in ~~`device_state_attributes()` where it belongs.~~ in a utility method called from `update()`, for code clarity and since there is potential I/O from SQLAlchemy.
- As a bonus, many metadata entries with cryptic values have complementary entries added that provide easier to use data:
	- .\* Stop Drop Off Type: .\* Stop Drop Off Type **State** -> (string, unknown)
	- .\* Stop Pickup Type: .\* Stop Pickup Type **State** -> (string, unknown)
	- .\* Stop Timepoint: .\* Stop Timepoint **Exact** -> boolean
	- .\* Station Location Type: .\* Station Location Type **Name** -> string
	- .\* Wheelchair Boarding: .\* Wheelchair Boarding **Available** -> (boolean, unknown)
	- Route Type: Route Type **Name** (string)
	- Trip Bikes Allowed: Trip Bikes Allowed **State** -> (boolean, unknown)
	- Trip Wheelchair Access: Trip Wheelchair Access **Available** -> (boolean, unknown)
- Attribute names are now using snake_case.
- Added type hints.

**Related issue (if applicable):** fixes #21222

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

dilruacs committed:
Modify check for ADB public key (#22378)

* Remove check for public key

* Remove has_adb_files, directly call cv.isfile

* Check for missing adbkey.pub, create dummy if not found

* Reorder imports

* Bumped androidtv library version, deactivated pubkey test

* Code works without pubkey, removed function

* Removed "import os", not needed anymore

* Bump library version

rohankapoorcom committed:
Centralize all usages of `host` as a configuration param to the constant CONF_HOST (#22458)

* Centralize all usages of  as a configuration param to the constant CONF_HOST

* Clean up test

andrewsayre committed:
Don't return cover position when not supported (#22484)

Adminiuga committed:
Targeted ZHA permit joins. (#22482)

* Targeted ZHA permit service.

* Convert IEEE string to EUI64 usiv vol schema.

* Update test units.

* Lint.

isort imports.

syssi committed:
Fix auto discovery of yeelights (#22481)

* Fix auto discovery of yeelights

* Fix lint

heythisisnate committed:
fix inverse state changes for binary sensors (#22479)

Kane610 committed:
Fix events so they work with multiple devices (#22477)

Danielhiversen committed:
Improve Sensibo error handling (#22475)

* Handle sensibo exception

* improve sensibo

bieniu committed:
Add new data fields and bump python-join-api (#22472)

* Add new data fields and bump python-join-api

* Update __init__.py

fbradyirl committed:
Bump pypi again for Cisco Mobility Express (#22467)

* Fix empty device clId

Fix for https://github.com/fbradyirl/ciscomobilityexpress/issues/13

* Update requirements_all.txt

jackwilsdon committed:
Return percentage information in Alexa Smart Home response (#22440)

Jc2k committed:
Move HKDevice into connection (#22430)

amelchio committed:
Incoming SMS handling for netgear_lte (#22402)

* Fire netgear_lte events for incoming SMS

* Add netgear_lte.delete_sms service call

* Fix log statement

* Add services.yaml

hfurubotten committed:
Entur upgrade to v0.2.0: async polling, number of departures, omit non boarding departures (#22001)

* The Great Migration

* removed outdated tests

* Upgraded enturclient to v0.2.0
- client has gone async
- configurable if non boarding departures should be omitted from result
- configurable how many departures should be shown
- more clear which attributes are included and when

* reduced nesting and clearing of attributes

* Removed test data

* ensure attribution and stop id in attributes

* docstring fixes

* fix rebase errors
went a bit too fast on continue on one of the conflicts..

* fix test requirements comment from gen

rohankapoorcom committed:
Mopar split (#21526)

* Split out mopar into a component and sensor platform

* Add the mopar switch platform

* Add the mopar lock platform

* Clean up and bump version

* Update per review

* Re-add service to trigger horn

* Clean up again

* Don't call async from sync context

* Lint

* Implement changes from review

* Lint

* A little more clean up

bachya committed:
Fix too-abrubt SimpliSafe data refresh termination on error (#22466)

balloob committed:
Fix syntax error

robbiet480 committed:
Great migration notify (#22406)

* Move notify platforms into components

* Move notify tests

* Fix notify tests

* More fixes

* Update requirements

* Update .coveragerc

* Run gen reqs

balloob committed:
Print error instead of warning for custom platforms in legacy format (#22486)

* Legacy platform format prints error

* Enforce no partial overlays

cliffordwhansen committed:
Add game and app media types (#22459)

* Added game and app media types

* Changed media type to game from music

* Removed app type as it is not used yet

balloob committed:
Updated frontend to 20190327.0

balloob committed:
Update translations

hunterjm committed:
Stream Record Service (#22456)

* Initial commit of record service for live streams

* fix lint

* update service descriptions

* add tests

* fix lint

awarecan committed:
Add trusted_users in trusted networks auth provider (#22478)

balloob committed:
Merge remote-tracking branch 'origin/master' into dev

balloob committed:
Version bump to 0.92.0dev0

balloob committed:
Fix YAML

awarecan committed:
Fix dev branch (#22493)

awarecan committed:
A very basic Circleci setup (#22503)

* Add circleci support

* Add buildpack-deps

* Install libudev-dev

* sudo

* always run test

* Add test report

* no sugar

* quite pytest

* better junit test result

* Add $CODE_COVERAGE env var

awarecan committed:
Fix Circleci config (#22509)

* Add libav depends on circleci

* tweak circleci config

awarecan committed:
Remove botocore dependency from credstash script (#22511)

* Remove botocore dependency from credstash script

* Update requirements_all.txt

* Update pylintrc

* Update credstash.py

alengwenus committed:
Add LCN binary_sensor component (#22341)

awarecan committed:
Fix lint on dev (#22512)

## Description:

Fix a lint issue in credstash script.


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

MartinHjelmare commented on homeassistant/components/ads/binary_sensor.py:
```diff
@@ -41,22 +44,36 @@ def __init__(self, ads_hub, name, ads_var, device_class):
             self._ads_hub.add_device_notification,
             self.ads_var, self._ads_hub.PLCTYPE_BOOL, update)
+        try:
+            with async_timeout.timeout(30):
```
FYI: This is longer than the home assistant timeout for logging a warning for slow platform setup, which is 10 seconds.

MartinHjelmare commented on homeassistant/components/ads/switch.py:
```diff
@@ -29,7 +32,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    add_entities([AdsSwitch(ads_hub, name, ads_var)])
 
 
 class AdsSwitch(ToggleEntity):
```
It would be nice if we refactor and extract the common entity logic into a common base ADS entity class, and avoid the copy paste. We don't need to do that in this PR.

Looks like we can extract:
`async_added_to_hass`, `is_on`, `name`, `unique_id`, `should_poll`, and `available` if we rename some attributes and move out some functions that have platform specific names and logic.

MartinHjelmare reviewed (approved):
I think this is ok.

MartinHjelmare commented:
Please rebase on latest dev so that the build checks can complete.

easink committed:
Add HEOS media player component (#21721)

## Description:
Denon HEOS media player.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8848

## Example entry for `configuration.yaml` (if applicable):
```yaml
heos:
  host: HEOS-1                              
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23


Co-authored-by: Andrew Sayre <6730289+andrewsayre@users.noreply.github.com>

yosilevy committed:
Add google calendar max_results config option (#21874)

* Added max_results config capability to google calendar (people are creating custom components just to override that)

* Dummy commit

* Dummy commit 2

* Changed to positive_int

* Removed double imports

andrewsayre committed:
Fix .coveragerc from merge/rebase (#22516)

* Fix coveragerc

* Fix coveragerc

carstenschroeder committed:
Merge branch 'ads-binary-sensor-fix' of https://github.com/carstenschroeder/home-assistant into ads-binary-sensor-fix

carstenschroeder commented:
Sorry, I messed it up completely. I will create a new clean PR.
